semantics/src/main/scala/aqua/semantics/rules/locations/LocationsState.scala

```diff
@@ -1,26 +1,58 @@
 package aqua.semantics.rules.locations
 
 import aqua.parser.lexer.Token
-import aqua.semantics.rules.types.TypesState
 import cats.kernel.Monoid
+import scribe.Logging
 
 case class LocationsState[S[_]](
-  tokens: Map[String, Token[S]] = Map.empty[String, Token[S]],
-  locations: List[(Token[S], Token[S])] = Nil,
-  stack: List[LocationsState[S]] = Nil
-) {
+  variables: List[VariableInfo[S]] = Nil
+) extends Logging {
 
-  lazy val allLocations: List[(Token[S], Token[S])] = locations
+  def addDefinitions(newDefinitions: List[DefinitionInfo[S]]): LocationsState[S] =
+    copy(variables = newDefinitions.map(d => VariableInfo(d)) ++ variables)
+
+  def addDefinition(newDef: DefinitionInfo[S]): LocationsState[S] =
+    copy(variables = VariableInfo(newDef) +: variables)
+
+  private def addOccurrenceToFirst(
+    vars: List[VariableInfo[S]],
+    name: String,
+    token: Token[S]
+  ): List[VariableInfo[S]] = vars match {
+    case Nil =>
+      logger.error(s"Unexpected. Cannot add occurrence for $name")
+      Nil
+    case head :: tail =>
+      if (head.definition.name == name)
+        head.copy(occurrences = token +: head.occurrences) :: tail
+      else
+        head :: addOccurrenceToFirst(tail, name, token)
+  }
+
+  def addLocation(
+    name: String,
+    token: Token[S]
+  ): LocationsState[S] = {
+    copy(variables = addOccurrenceToFirst(variables, name, token))
+  }
+
+  def addLocations(
+    locations: List[(String, Token[S])]
+  ): LocationsState[S] =
+    locations.foldLeft(this) { case (st, (name, token)) =>
+      st.copy(variables = addOccurrenceToFirst(variables, name, token))
```

Comment:
Do I understand right that `variables` referenced here are `variables` of outer state? So we add only the last location here because state is actually not saved. You can use `addLocation` here instead of repeating it and issue will be fixed